zipkin-storage/elasticsearch-rest/src/main/java/zipkin/storage/elasticsearch/ElasticsearchRestSpanConsumer.java

```diff
@@ -0,0 +1,78 @@
+/**
+ * Copyright 2015-2016 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package zipkin.storage.elasticsearch;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import org.apache.http.HttpEntity;
+import org.apache.http.entity.ContentType;
+import org.apache.http.nio.entity.NStringEntity;
+import org.elasticsearch.client.Response;
+import org.elasticsearch.client.ResponseListener;
+import org.elasticsearch.client.RestClient;
+import zipkin.Codec;
+import zipkin.Span;
+import zipkin.internal.Util;
+import zipkin.storage.AsyncSpanConsumer;
+import zipkin.storage.Callback;
+
+import static zipkin.internal.ApplyTimestampAndDuration.guessTimestamp;
+import static zipkin.storage.elasticsearch.ElasticsearchSpanConsumer.prefixWithTimestampMillis;
+
+final class ElasticsearchRestSpanConsumer implements AsyncSpanConsumer {
+  final RestClient client;
+  final IndexNameFormatter indexNameFormatter;
+
+  ElasticsearchRestSpanConsumer(RestClient client, IndexNameFormatter indexNameFormatter) {
+    this.client = client;
+    this.indexNameFormatter = indexNameFormatter;
+  }
+
+  @Override public void accept(List<Span> spans, final Callback<Void> callback) {
+    StringBuilder request = new StringBuilder();
+    for (Span s : spans) {
+      createSpanIndexRequest(s, request);
+    }
+
+    HttpEntity entity = new NStringEntity(request.toString(), ContentType.APPLICATION_JSON);
+    client.performRequest("POST", "/_bulk", Collections.<String, String>emptyMap(), entity,
```

Comment:
luckily you are talking to the original author of jclouds.. I think I've
written that signature code several times :P

On Wed, Sep 14, 2016 at 12:24 AM, Brian Devins <notifications@github.com>
wrote:

> In zipkin-storage/elasticsearch-rest/src/main/java/zipkin/
> storage/elasticsearch/ElasticsearchRestSpanConsumer.java
> <https://github.com/openzipkin/zipkin/pull/1294#discussion_r78594416>:
>
> > +  final RestClient client;
> > +  final IndexNameFormatter indexNameFormatter;
> > +
> > +  ElasticsearchRestSpanConsumer(RestClient client, IndexNameFormatter indexNameFormatter) {
> > +    this.client = client;
> > +    this.indexNameFormatter = indexNameFormatter;
> > +  }
> > +
> > +  @Override public void accept(List<Span> spans, final Callback<Void> callback) {
> > +    StringBuilder request = new StringBuilder();
> > +    for (Span s : spans) {
> > +      createSpanIndexRequest(s, request);
> > +    }
> > +
> > +    HttpEntity entity = new NStringEntity(request.toString(), ContentType.APPLICATION_JSON);
> > +    client.performRequest("POST", "/_bulk", Collections.<String, String>emptyMap(), entity,
>
> In order to authenticate requests to amazon when locking down the security
> on our domain, we will need to be able to add on an authentication header
> which is a signature based on the request body.
>
> http://docs.aws.amazon.com/elasticsearch-service/latest/
> developerguide/what-is-amazon-elasticsearch-service.html#signing-requests
>
> http://docs.aws.amazon.com/general/latest/gr/signature-
> v4-examples.html#signature-v4-examples-java
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/openzipkin/zipkin/pull/1294/files/f6fcf14595940b112d303c7105a0ca126de963a8#r78594416>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAD61-Wpzl0RYsZkSIsNuZ0PPrMxjndTks5qps4zgaJpZM4J7dnf>
> .
>